zkvm/src/predicate.rs

```diff
@@ -136,42 +105,43 @@ impl Predicate {
         Predicate::Opaque(self.to_point())
     }
 
-    /// Creates a disjunction of a vector of predicates.
-    pub fn disjunction(preds: Vec<Predicate>) -> Result<Self, VMError> {
-        let point = {
-            let f = Predicate::commit_disjunction(preds.iter().map(|p| p.to_point()));
-            let l = preds[0]
-                .to_point()
-                .decompress()
-                .ok_or(VMError::InvalidPoint)?;
-            l + f * PedersenGens::default().B
-        };
-        Ok(Predicate::Or(PredicateDisjunction {
-            preds: preds,
-            precomputed_point: point.compress(),
-        }))
-    }
-
-    fn commit_disjunction<I>(preds: I) -> Scalar
-    where
-        I: IntoIterator,
-        I::Item: Borrow<CompressedRistretto>,
-        I::IntoIter: ExactSizeIterator,
-    {
-        let mut t = Transcript::new(b"ZkVM.predicate");
-        let iter = preds.into_iter();
-        t.commit_u64(b"n", iter.len() as u64);
-        for x in iter {
-            t.commit_point(b"X", x.borrow());
+    fn commit_taproot(key: &[u8], root: &[u8]) -> Scalar {
+        let mut t = Transcript::new(b"ZkVM.taproot");
+        t.commit_bytes(b"key", key);
+        t.commit_bytes(b"merkle", root);
+        t.challenge_scalar(b"h")
+    }
+
+    /// Verifies whether the current predicate is a commitment to a signing key `key` and Merkle root `root`.
+    /// Returns a `PointOp` instance that can be verified in a batch with other operations.
+    pub fn prove_taproot(&self, program: &Program, call_proof: &CallProof) -> PointOp {
```

Comment:
This method cannot take `Program`, it must take a bytecode — the Verifier is not going to parse the program only for it to be encoded for verification, and then parsed again during execution. 

```suggestion
    pub fn prove_taproot(&self, program: &[u8], call_proof: &CallProof) -> PointOp {
```

Reply:
This makes sense to me in theory. However, the only use of `program` in this function is as an argument to `compute_root_from_path`. That particular argument is an `Entry`, which requires the implementation of `MerkleItem`. This doesn't seem possible for a byte-array. This seems like it might require some larger changes.. what do you think?